aspnet/mvc/models/validation.rst

```diff
@@ -1,8 +1,125 @@
+- ``[ForeignKey]``: Provides foreign key functionality.
+- ``[Key]``: Provides primary key functionality. 
+- ``[Phone]``: Formats the property as a telephone format.
+- ``[Range]``: Sets the minimum and maximum values for multiple types of properties.
+- ``[RegularExpression]``: Forces the data to match a regular expression.
+- ``[Required]``: Makes a property required.
+- ``[ScaffoldColumn]``: Determines whether to render the property it is applied to.
+- ``[StringLength]``: The maximum length for a string field.
+- ``[Url]``: Formats the property as a URL.
+
+A complete list of annotations is at the `System.ComponentModel.DataAnnotations <https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations(v=vs.110).aspx>`_ namespace docs. Note that this namespace is not specific to MVC, but in a shared .NET namespace. 
+
+Many annotations provide actions similar to constraint checking in database tables, or in popular business logic patterns. However, there are intances where business rules don't fit into these common scenarios. For those times, you can create custom validation attributes to apply to model properties.  
+
+Model State
+-----------
+Model state gives you insight as to which form fields have not passed validation. Controllers contain a ``ModelState`` property that is of type ``ModelStateEntry`` and it contains an ``Errors`` property. The errors are a list of keys and values that contains the error messages and other information. You can use the error information to build a response so that the client will know which fields did not pass validation. In the case of a validation error, APIs should return an HTTP status code ``400 Bad Request``.
+
+If you have some logic that can produce a validation error, add that error to the model state's collection of errors by calling ``AddModelStateError`` as demonstrated below. Doing so will return the error message to the client and the view will display it.
+
+.. literalinclude:: validation/sample/MoviesController.cs
+   :language: c#
+   :lines: 48-69   
+   :emphasize-lines: 52-59
+   :dedent: 8
+   
+After model binding and validation are complete, you may want to repeat parts of it. For example, a user may have entered text in a field expecting an integer. You can invoke model binding to update the values of the model by using the ``UpdateModel`` or ``TryUpdateModel`` methods. When using ``UpdateModel``, you must put your code in `try-catch` block. It will attempt to update the model, but if it fails, it will throw an ``InvalidOperationException`` you must catch. ``TryUpdateModel`` attempts to update the model, but comes with a built-in exception handling in case it fails. Both methods will attempt to update the model state with validation errors, and then your view will display those errors. There is a similar pair of methods for validation called ``ValidateModel`` and ``TryValidateModel``. They work in the same manner but attempt to perform validation instead. 
```

Comment:
- `TryUpdateModel` and `UpdateModel` don't exist; think you mean `TryUpdateModelAsync` and there's no need to catch an `InvalidOperationException`. Similarly `TryValidateModel` exists but `ValidateModel` does not.
- More generally `TryUpdateModelAsync` doesn't seem to fit in this document. Suggest talking only about `TryValidateModel` here.